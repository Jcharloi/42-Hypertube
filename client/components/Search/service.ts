import qs from "qs";

export const clamp = (str: string, maxLength: number): string =>
  str
    ? `${str.slice(0, maxLength)}${str.length <= maxLength ? "" : "..."}`
    : "";

export const formatQueryUrl = (
  search: string,
  page: number,
  mediaType: string
): string => {
  const { query, collections, year, minRating } = qs.parse(search.slice(1));

  return `/${mediaType}?${qs.stringify({
    query: query || undefined,
    collections: collections || undefined,
    year: year || undefined,
    minRating: minRating === 0 ? undefined : minRating,
    page
  })}`;
};
