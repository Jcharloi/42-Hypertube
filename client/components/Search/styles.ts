import { makeStyles } from "@material-ui/core/styles";
import { yellow } from "@material-ui/core/colors";

const useSearchStyles = makeStyles((theme) => ({
  container: {
    display: "flex",
    flexWrap: "wrap",
    justifyContent: "center"
  },
  thumbnailContainer: {
    position: "relative",
    height: "450px",
    width: "300px",
    margin: theme.spacing(5),
    transition: "transform 0.2s",
    "&:hover": {
      [theme.breakpoints.up("sm")]: {
        transform: "scale(1.15)"
      }
    }
  },
  thumbnailOverlay: {
    textAlign: "center",
    position: "absolute",
    display: "flex",
    flexDirection: "column",
    top: 0,
    height: "100%",
    width: "100%",
    padding: theme.spacing(2, 1),
    alignItems: "center",
    transition: "background 0.2s, opacity 0.2s, transform 0.2s",
    opacity: 0,
    "&:hover": {
      opacity: 1,
      background: "rgba(0,0,0,0.8)"
    }
  },
  skeletonThumbnail: {
    background: theme.palette.secondary.main
  },
  skeletonProgress: {
    position: "absolute",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    top: 0,
    height: "100%",
    width: "100%",
    margin: theme.spacing(1, 0)
  },
  tag: {
    margin: theme.spacing(0.5)
  },
  summary: {
    margin: theme.spacing(2, 0),
    padding: theme.spacing(0, 1),
    flexGrow: 1,
    overflow: "auto",
    textAlign: "justify"
  },
  metaInfos: {
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    width: "100%"
  },
  rating: {
    display: "flex",
    marginBottom: theme.spacing(1)
  },
  watchButton: {
    width: "75%"
  },
  ratingIcon: {
    color: yellow[500]
  },
  altContainer: {
    height: 450,
    width: 300,
    padding: theme.spacing(2),
    display: "flex",
    justifyContent: "center",
    alignItems: "center"
  },
  noMediaContainer: {
    textAlign: "center",
    marginTop: theme.spacing(5)
  },
  noMedia: {
    width: "20rem",
    height: "20rem"
  }
}));

export default useSearchStyles;
