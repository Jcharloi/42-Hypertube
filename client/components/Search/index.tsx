import React, { ReactElement, useState, useEffect } from "react";
import { useHistory, useLocation } from "react-router-dom";
import { useIntl } from "react-intl";
import _ from "lodash";
import InfiniteScroll from "react-infinite-scroller";

import Skeleton from "@material-ui/lab/Skeleton";
import CircularProgress from "@material-ui/core/CircularProgress";
import Container from "@material-ui/core/Container";
import Typography from "@material-ui/core/Typography";
import StarIcon from "@material-ui/icons/Star";
import { Button, Chip } from "@material-ui/core";
import Image from "material-ui-image";
import Card from "@material-ui/core/Card";

import useApi from "../../hooks/useApi";
import { formatQueryUrl } from "./service";
import { SearchData } from "../../models/models";

import useSearchStyles from "./styles";

const Search = (): ReactElement => {
  const classes = useSearchStyles({});
  const { formatMessage: _t } = useIntl();
  const [filmList, setFilmList] = useState([]);
  const [block, setBlock] = useState(false);
  const [page, setPage] = useState(1);
  const history = useHistory();
  const location = useLocation();
  const mediaType = location.pathname.includes("movies") ? "movies" : "shows";
  const { data, loading, error, setUrl }: SearchData = (useApi(
    formatQueryUrl(location.search, 1, mediaType)
  ) as unknown) as SearchData;

  useEffect(() => {
    if (!loading && !block) {
      setPage(1);
      setFilmList([]);
      setUrl(formatQueryUrl(location.search, 1, mediaType));
    }
  }, [location.search, location.pathname]);

  useEffect(() => {
    if (data && data.medias?.length) {
      setFilmList(() => [...filmList, ...data.medias]);
      setBlock(false);
    }
  }, [data]);

  const loadMore = (): void => {
    if (!data.nextPage) {
      return;
    }
    setBlock(true);
    setPage(page + 1);
    setUrl(formatQueryUrl(location.search, page + 1, mediaType));
  };

  if (error) {
    history.push("/error");
  }

  return (
    <Container maxWidth="xl">
      <InfiniteScroll
        className={classes.container}
        initialLoad={false}
        loadMore={loadMore}
        hasMore={data.nextPage}
      >
        {!filmList.length && !loading && (
          <div className={classes.noMediaContainer}>
            <img
              src="http://localhost:8080/public/no-media.png"
              alt="no-media"
              className={classes.noMedia}
            />
            <Typography variant="h5">
              {_t({ id: "search.no_media" })}
            </Typography>
          </div>
        )}
        {filmList.map((media) => (
          <div
            className={classes.thumbnailContainer}
            key={_.uniqueId("media_")}
          >
            <Image
              animationDuration={500}
              src={media.cover}
              color="rgba(0,0,0,0)"
              imageStyle={{ width: 300, height: 450 }}
              style={{ width: 300, height: 450 }}
              disableSpinner
              errorIcon={
                <Card className={classes.altContainer}>
                  <Typography variant="h4">{media.title}</Typography>
                </Card>
              }
            />
            <div className={classes.thumbnailOverlay}>
              <Typography variant="h5">{media.title}</Typography>
              <Typography variant="h6">{media.year}</Typography>
              <Typography variant="caption" className={classes.summary}>
                {media.summary}
              </Typography>
              <div className={classes.metaInfos}>
                {media.genres && (
                  <div>
                    {media.genres.map((genre: string) => (
                      <Chip
                        className={classes.tag}
                        label={genre}
                        clickable
                        color="primary"
                        key={_.uniqueId("mediaGenre_")}
                      />
                    ))}
                  </div>
                )}
                <Typography className={classes.rating}>
                  {media.rating} <StarIcon className={classes.ratingIcon} />
                  {media.runtime && <span>- {media.runtime} mins</span>}
                </Typography>
                <Button
                  color="primary"
                  variant="contained"
                  className={classes.watchButton}
                  onClick={(): void =>
                    history.push(`/${mediaType.slice(0, -1)}/${media.id}`)
                  }
                >
                  {_t({
                    id:
                      mediaType === "movies"
                        ? "search.film.watch"
                        : "search.film.watch_show"
                  })}
                </Button>
              </div>
            </div>
          </div>
        ))}
        {loading &&
          _.times(12, () => (
            <div
              className={classes.thumbnailContainer}
              key={_.uniqueId("loading_")}
            >
              <Skeleton
                variant="rect"
                width={300}
                height={450}
                className={classes.skeletonThumbnail}
              />
              <div className={classes.skeletonProgress}>
                <CircularProgress />
              </div>
            </div>
          ))}
      </InfiniteScroll>
    </Container>
  );
};

export default Search;
