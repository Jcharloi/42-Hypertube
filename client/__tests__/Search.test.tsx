import React from "react";
import Adapter from "enzyme-adapter-react-16";
import EnzymeToJson from "enzyme-to-json";
import { configure } from "enzyme";

import Search from "../components/Search";

import { mountWithIntl } from "./helpers/intl-enzyme-test-helper";

import * as ServiceSearch from "../components/Search/service";

configure({ adapter: new Adapter() });

jest.mock("react-router-dom", () => ({
  useHistory: (): Record<string, unknown> => ({
    push: jest.fn(),
    location: { state: {}, pathname: "/search/movies" }
  }),
  useLocation: (): Record<string, unknown> => ({
    pathname: "/search/movies",
    state: {},
    search: ""
  })
}));

jest.mock("../hooks/useApi", () => (): {
  data: unknown;
  loading: boolean;
  error: void;
  setUrl: () => void;
} => ({
  data: [],
  loading: false,
  error: null,
  setUrl: jest.fn()
}));

describe("Search", () => {
  describe("search en", () => {
    it("renders correctly", () => {
      const domNode = mountWithIntl(<Search />, "en");
      expect(EnzymeToJson(domNode)).toMatchSnapshot();
    });
  });

  describe("search fr", () => {
    it("renders correctly", () => {
      const domNode = mountWithIntl(<Search />, "fr");
      expect(EnzymeToJson(domNode)).toMatchSnapshot();
    });
  });

  describe("Service", () => {
    describe("clamp", () => {
      it("should not clamp", () => {
        const shortFilmName = "Short";

        const clamped = ServiceSearch.clamp(shortFilmName, 24);
        expect(clamped).toBe("Short");
      });

      it("should clamp", () => {
        const longFilmName = "Very long";

        const clamped = ServiceSearch.clamp(longFilmName, 1);
        expect(clamped).toBe("V...");
      });
    });

    describe("formatQueryUrl", () => {
      it("should get a full query", () => {
        const testQuery = ServiceSearch.formatQueryUrl(
          "?query=top&collections%5B0%5D=collec1&collections%5B1%5D=collec2",
          1,
          "movies"
        );

        expect(testQuery).toBe(
          "/movies?query=top&collections%5B0%5D=collec1&collections%5B1%5D=collec2&page=1"
        );
      });

      it("should get a empty query", () => {
        const testQuery = ServiceSearch.formatQueryUrl("", 1, "movies");
        expect(testQuery).toBe("/movies?page=1");
      });
    });
  });
});
