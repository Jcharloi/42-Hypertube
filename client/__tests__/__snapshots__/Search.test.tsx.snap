// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search search en renders correctly 1`] = `
<Search>
  <WithStyles(ForwardRef(Container))
    maxWidth="xl"
  >
    <ForwardRef(Container)
      classes={
        Object {
          "disableGutters": "MuiContainer-disableGutters",
          "fixed": "MuiContainer-fixed",
          "maxWidthLg": "MuiContainer-maxWidthLg",
          "maxWidthMd": "MuiContainer-maxWidthMd",
          "maxWidthSm": "MuiContainer-maxWidthSm",
          "maxWidthXl": "MuiContainer-maxWidthXl",
          "maxWidthXs": "MuiContainer-maxWidthXs",
          "root": "MuiContainer-root",
        }
      }
      maxWidth="xl"
    >
      <div
        className="MuiContainer-root MuiContainer-maxWidthXl"
      >
        <InfiniteScroll
          className="makeStyles-container-1"
          element="div"
          getScrollParent={null}
          hasMore={false}
          initialLoad={false}
          isReverse={false}
          loadMore={[Function]}
          loader={null}
          pageStart={0}
          ref={null}
          threshold={250}
          useCapture={false}
          useWindow={true}
        >
          <div
            className="makeStyles-container-1"
          >
            <div
              className="makeStyles-noMediaContainer-13"
            >
              <img
                alt="no-media"
                className="makeStyles-noMedia-14"
                src="http://localhost:8080/public/no-media.png"
              />
              <WithStyles(ForwardRef(Typography))
                variant="h5"
              >
                <ForwardRef(Typography)
                  classes={
                    Object {
                      "alignCenter": "MuiTypography-alignCenter",
                      "alignJustify": "MuiTypography-alignJustify",
                      "alignLeft": "MuiTypography-alignLeft",
                      "alignRight": "MuiTypography-alignRight",
                      "body1": "MuiTypography-body1",
                      "body2": "MuiTypography-body2",
                      "button": "MuiTypography-button",
                      "caption": "MuiTypography-caption",
                      "colorError": "MuiTypography-colorError",
                      "colorInherit": "MuiTypography-colorInherit",
                      "colorPrimary": "MuiTypography-colorPrimary",
                      "colorSecondary": "MuiTypography-colorSecondary",
                      "colorTextPrimary": "MuiTypography-colorTextPrimary",
                      "colorTextSecondary": "MuiTypography-colorTextSecondary",
                      "displayBlock": "MuiTypography-displayBlock",
                      "displayInline": "MuiTypography-displayInline",
                      "gutterBottom": "MuiTypography-gutterBottom",
                      "h1": "MuiTypography-h1",
                      "h2": "MuiTypography-h2",
                      "h3": "MuiTypography-h3",
                      "h4": "MuiTypography-h4",
                      "h5": "MuiTypography-h5",
                      "h6": "MuiTypography-h6",
                      "noWrap": "MuiTypography-noWrap",
                      "overline": "MuiTypography-overline",
                      "paragraph": "MuiTypography-paragraph",
                      "root": "MuiTypography-root",
                      "srOnly": "MuiTypography-srOnly",
                      "subtitle1": "MuiTypography-subtitle1",
                      "subtitle2": "MuiTypography-subtitle2",
                    }
                  }
                  variant="h5"
                >
                  <h5
                    className="MuiTypography-root MuiTypography-h5"
                  >
                    No medias found !
                  </h5>
                </ForwardRef(Typography)>
              </WithStyles(ForwardRef(Typography))>
            </div>
          </div>
        </InfiniteScroll>
      </div>
    </ForwardRef(Container)>
  </WithStyles(ForwardRef(Container))>
</Search>
`;

exports[`Search search fr renders correctly 1`] = `
<Search>
  <WithStyles(ForwardRef(Container))
    maxWidth="xl"
  >
    <ForwardRef(Container)
      classes={
        Object {
          "disableGutters": "MuiContainer-disableGutters",
          "fixed": "MuiContainer-fixed",
          "maxWidthLg": "MuiContainer-maxWidthLg",
          "maxWidthMd": "MuiContainer-maxWidthMd",
          "maxWidthSm": "MuiContainer-maxWidthSm",
          "maxWidthXl": "MuiContainer-maxWidthXl",
          "maxWidthXs": "MuiContainer-maxWidthXs",
          "root": "MuiContainer-root",
        }
      }
      maxWidth="xl"
    >
      <div
        className="MuiContainer-root MuiContainer-maxWidthXl"
      >
        <InfiniteScroll
          className="makeStyles-container-1"
          element="div"
          getScrollParent={null}
          hasMore={false}
          initialLoad={false}
          isReverse={false}
          loadMore={[Function]}
          loader={null}
          pageStart={0}
          ref={null}
          threshold={250}
          useCapture={false}
          useWindow={true}
        >
          <div
            className="makeStyles-container-1"
          >
            <div
              className="makeStyles-noMediaContainer-13"
            >
              <img
                alt="no-media"
                className="makeStyles-noMedia-14"
                src="http://localhost:8080/public/no-media.png"
              />
              <WithStyles(ForwardRef(Typography))
                variant="h5"
              >
                <ForwardRef(Typography)
                  classes={
                    Object {
                      "alignCenter": "MuiTypography-alignCenter",
                      "alignJustify": "MuiTypography-alignJustify",
                      "alignLeft": "MuiTypography-alignLeft",
                      "alignRight": "MuiTypography-alignRight",
                      "body1": "MuiTypography-body1",
                      "body2": "MuiTypography-body2",
                      "button": "MuiTypography-button",
                      "caption": "MuiTypography-caption",
                      "colorError": "MuiTypography-colorError",
                      "colorInherit": "MuiTypography-colorInherit",
                      "colorPrimary": "MuiTypography-colorPrimary",
                      "colorSecondary": "MuiTypography-colorSecondary",
                      "colorTextPrimary": "MuiTypography-colorTextPrimary",
                      "colorTextSecondary": "MuiTypography-colorTextSecondary",
                      "displayBlock": "MuiTypography-displayBlock",
                      "displayInline": "MuiTypography-displayInline",
                      "gutterBottom": "MuiTypography-gutterBottom",
                      "h1": "MuiTypography-h1",
                      "h2": "MuiTypography-h2",
                      "h3": "MuiTypography-h3",
                      "h4": "MuiTypography-h4",
                      "h5": "MuiTypography-h5",
                      "h6": "MuiTypography-h6",
                      "noWrap": "MuiTypography-noWrap",
                      "overline": "MuiTypography-overline",
                      "paragraph": "MuiTypography-paragraph",
                      "root": "MuiTypography-root",
                      "srOnly": "MuiTypography-srOnly",
                      "subtitle1": "MuiTypography-subtitle1",
                      "subtitle2": "MuiTypography-subtitle2",
                    }
                  }
                  variant="h5"
                >
                  <h5
                    className="MuiTypography-root MuiTypography-h5"
                  >
                    Aucun média trouvé !
                  </h5>
                </ForwardRef(Typography)>
              </WithStyles(ForwardRef(Typography))>
            </div>
          </div>
        </InfiniteScroll>
      </div>
    </ForwardRef(Container)>
  </WithStyles(ForwardRef(Container))>
</Search>
`;
